### PR TITLE
Chiwawasm

### PR DESCRIPTION
**Pull request by lydia-pierce**
Added cosm wasm to app.go based on the [Juno code](https://github.com/CosmosContracts/juno/blob/d323777476c68c31bd906492c223fc42acf70e30/app/app.go)

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -446,6 +550,20 @@ func New(
...
+	// 	panic(err)
+	// }
+
+	// if upgradeInfo.Name == "moneta" && !app.UpgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
+	// 	// storeUpgrades := store.StoreUpgrades{
+	// 	// 	Added: []string{authz.ModuleName, feegrant.ModuleName, wasm.ModuleName},
+	// 	// }
+
+	// 	// // configure store loader that checks if version == upgradeHeight and applies store upgrades
+	// 	// app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
```
Comment (lydia-pierce) [inline app/app.go]: I was unsure if this code was necessary, or if it was just specific to the [Juno moneta](https://github.com/CosmosContracts/juno/issues/87) upgrade, so I left it in but commented it out. If it is unnecessary, we can delete it.

Comment (evan-forbes) [reply app/app.go]: we might be able to use it to upgrade, iirc juno did the same, launched a chain and then upgraded to using smart contracts

File: app/ante.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+			if msg.CommissionRate == nil {
+				break
+			}
+			if msg.CommissionRate.LT(minCommissionRate) {
+				return sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "commission can't be lower than 5%")
+			}
+		}
+
+		return nil
+	}
```
Comment (evan-forbes) [inline app/ante.go]: we don't need this entire ante handler, as we have min commission implemented in a different way in the sdk

**Comment by evan-forbes** [comment]
do you know if we need https://github.com/CosmosContracts/juno/blob/v2.1.0/app/wasm_config.go ?

also, let's grab the cli command https://github.com/CosmosContracts/juno/blob/main/cmd/junod/genwasm.go

File: app/app.go [hunk 3]
```diff
@@ -88,25 +93,72 @@ import (
...
+	return proposals
+}
+
+func GetWasmOpts(appOpts servertypes.AppOptions) []wasm.Option {
+	var wasmOpts []wasm.Option
+	if cast.ToBool(appOpts.Get("telemetry.enabled")) {
+		wasmOpts = append(wasmOpts, wasmkeeper.WithVMCacheMetrics(prometheus.DefaultRegisterer))
+	}
+
+	wasmOpts = append(wasmOpts, wasmkeeper.WithGasRegister(NewChihuahuaWasmGasRegister()))
```
Comment (lydia-pierce) [inline app/app.go:143]: > do you know if we need https://github.com/CosmosContracts/juno/blob/v2.1.0/app/wasm_config.go ?
> 

This function `NewChihuahuaWasmGasRegister()` which comes from wasm_config.go, is used here in app.go, and is used in the definition of `wasmOpts` (line 406) which is used in the `app.wasmKeeper` (line 407). Afaict it seems like this function is pretty integrated and necessary.

File: app/ante.go (hunk 2 above)
Comment (lydia-pierce) [reply app/ante.go]: Did you mean that we don't need this entire file? Or just the `AnteHandle()` function.

File: app/app.go (hunk 1 above)
Comment (joeabbey) [reply app/app.go]: yah.. it's necessary... we need to produce new merkle trees for new modules...

(see https://github.com/osmosis-labs/osmosis/pull/610)

specifically
```
	upgradeInfo, err := app.UpgradeKeeper.ReadUpgradeInfoFromDisk()
	if err != nil {
		panic(fmt.Sprintf("failed to read upgrade info from disk %s", err))
	}

	if upgradeInfo.Name == v5UpgradeName && !app.UpgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
		storeUpgrades := store.StoreUpgrades{
			Added: []string{authz.ModuleName, txfees.ModuleName},
		}

		// configure store loader that checks if version == upgradeHeight and applies store upgrades
		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
	}
```

File: app/ante.go (hunk 2 above)
Comment (evan-forbes) [reply app/ante.go]: we just don't need anything related to enforcing the minimum commission, so just `MinCommissionDecorator`

**Pull request merged**